File: Payload_Type/poseidon/mythic/agent_functions/dyld_inject.py
from mythic_payloadtype_container.MythicCommandBase import *
import json

class DyldInjectArguments(TaskArguments):
    def __init__(self, command_line, **kwargs):
        super().__init__(command_line, **kwargs)
        self.args = [
             CommandParameter(
                name="application",
                type=ParameterType.String,
                description="Path to the target application/binary",
                parameter_group_info=[
                    ParameterGroupInfo(
                        required=True,
                        ui_position=1
                    )
                ]
            ),
            CommandParameter(
                name="dylibpath",
                type=ParameterType.String,
                description="Path to the dylib on disk that will be injected into the target application",
                parameter_group_info=[
                    ParameterGroupInfo(
                        required=True,
                        ui_position=2
                    )
                ]
            ),
            CommandParameter(
                name="hideApp",
                type=ParameterType.Boolean,
                default_value=False,
                description="If true, launch the application with the kLSLaunchAndHide flag set. If false, use the kLSLaunchDefaults flag",
                parameter_group_info=[
                    ParameterGroupInfo(
                        required=True,
                        ui_position=3
                    )
                ]
            )
        ]

    async def parse_arguments(self):
        if len(self.command_line) > 0:
            if self.command_line[0] == "{":
                self.load_args_from_json_string(self.command_line)
            else:
                raise ValueError("Missing JSON arguments")

        else:
            raise ValueError("Missing arguments")

    async def parse_dictionary(self, dictionary):
        self.load_args_from_dictionary(dictionary)
    
class DyldInjectCommand(CommandBase):
    cmd = "dyld_inject"
    needs_admin = False
    help_cmd = "dyld_inject"
    description = "Spawn an application/binary and inject a dylib into application with the DYLD_INSERT_LIBRARIES environment variable"
    version = 1
    author = "@xorrior @_r3ggi"
    attackmapping = ["T1574.006"]
    argument_class = DyldInjectArguments
    attributes = CommandAttributes(
        supported_os=[SupportedOS.MacOS]
    )

    async def create_tasking(self, task: MythicTask) -> MythicTask:
        return task

    async def process_response(self, response: AgentResponse):
        pass

File: Payload_Type/poseidon/poseidon/agentfunctions/dyld_inject.go
package agentfunctions

import (
	agentstructs "github.com/MythicMeta/MythicContainer/agent_structs"
)

func init() {
	agentstructs.AllPayloadData.Get("poseidon").AddCommand(agentstructs.Command{
		Name:                "dyld_inject",
		HelpString:          "dyld_inject",
		Description:         "Spawn an application/binary and inject a dylib into application with the DYLD_INSERT_LIBRARIES environment variable",
		Version:             1,
		MitreAttackMappings: []string{"T1574.006"},
		Author:              "@xorrior @_r3ggi",
		CommandParameters: []agentstructs.CommandParameter{
			{
				Name:          "application",
				ParameterType: agentstructs.COMMAND_PARAMETER_TYPE_STRING,
				Description:   "Path to the target application/binary",
				ParameterGroupInformation: []agentstructs.ParameterGroupInfo{
					{
						ParameterIsRequired: true,
						UIModalPosition:     1,
					},
				},
			},
			{
				Name:             "dylibpath",
				ModalDisplayName: "Dylib Path",
				ParameterType:    agentstructs.COMMAND_PARAMETER_TYPE_STRING,
				Description:      "Path to the dylib on disk that will be injected into the target application",
				ParameterGroupInformation: []agentstructs.ParameterGroupInfo{
					{
						ParameterIsRequired: true,
						UIModalPosition:     2,
					},
				},
			},
			{
				Name:          "hideApp",
				ParameterType: agentstructs.COMMAND_PARAMETER_TYPE_BOOLEAN,
				DefaultValue:  false,
				Description:   "If true, launch the application with the kLSLaunchAndHide flag set. If false, use the kLSLaunchDefaults flag",
				ParameterGroupInformation: []agentstructs.ParameterGroupInfo{
					{
						ParameterIsRequired: false,
						UIModalPosition:     3,
					},
				},
			},
		},
		CommandAttributes: agentstructs.CommandAttribute{
			SupportedOS: []string{agentstructs.SUPPORTED_OS_MACOS},
		},
		TaskFunctionParseArgString: func(args *agentstructs.PTTaskMessageArgsData, input string) error {
			return args.LoadArgsFromJSONString(input)
		},
		TaskFunctionParseArgDictionary: func(args *agentstructs.PTTaskMessageArgsData, input map[string]interface{}) error {
			return args.LoadArgsFromDictionary(input)
		},
		TaskFunctionCreateTasking: func(task *agentstructs.PTTaskMessageAllData) agentstructs.PTTaskCreateTaskingMessageResponse {
			response := agentstructs.PTTaskCreateTaskingMessageResponse{
				Success: true,
				TaskID:  task.Task.ID,
			}
			return response
		},
	})
}
